### ISSUE TITLE
bug(compiler) components containing moduleId crash webpack build - zone.js:260 Uncaught TypeError: uri.match is not a function

### ISSUE THREAD

**Issue by qdouble** [angular/angular#10626]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report
```
**Current behavior**
Any component containing a moduleId will cause a webpack build to crash, this is related to this bug I reported here https://github.com/angular/material2/issues/974 `url_resolver.js:238Uncaught TypeError: uri.match is not a function`

Having a moduleId like `moduleId = module.id` causes this function to crash the app as it will get passed a number instead of a string.

```
function _split(uri) {
    return uri.match(_splitRe);
}
```
my temp work around is to put this to stop the app from crashing:

```
function _split(uri) {
    if (typeof uri === 'string')
    return uri.match(_splitRe);
}
```

I'm not sure that this problem effects systemJS, but this bug should effect anyone who is using webpack.

**Comment by onedotover** [comment]
I had the same issue and found that converting the moduleId to a string before passing it to this function seemed to work ok as well. I'm not sure if that's a solution, though, because it doesn't make much sense searching for a URL from a number. Maybe moduleId has a new meaning in RC5?

Here is where it passes the ID in and the error is thrown:
https://github.com/angular/angular/blob/3d53b333919c81b292cb0f4b3e338a686f5a6a85/modules/%40angular/compiler/src/metadata_resolver.ts#L795

**Comment by kylecordes** [comment]
I experienced this problem and poked around and found the adjustment needed (remove all of these module ID lines added a month ago) this evening also. Ouch. I'm not particularly picky on whether or not that module line should be present, but if it needs to be removed and its presence will cause trouble, it would be extraordinarily kind of the team to make Angular emit some sort of helpful warning instead of this particular crash :-)

**Comment by qdouble** [comment]
it appears this pull request will fix this issue: https://github.com/angular/angular/pull/10595/commits/f54f9efc0e70eb1c71ac8eb6a8ed690f1e7f09f9

**Comment by JavascriptMick** [comment]
Does this mean that moduleId setting s redundant when using webpack builder?

**Comment by qdouble** [comment]
@JavascriptMick using a moduleId isn't necessary when using Webpack

however, this bug popped up with this commit (https://github.com/angular/angular/pull/10512/commits/a5b544e8736d4e0fa40744ca610b682fff2eec07) to the compiler when they removed the RegExpWrapper and which changed this: ` return RegExpWrapper.firstMatch(_splitRe, uri);` to `return uri.match(_splitRe);`

**Comment by kylecordes** [comment]
If the current behaviors to be believed, including the module in your component metadata is not only redundant, it is actively incorrect. It sounds like there is a PR that would make the module metadata be tolerated, but only barely so. If you trace through the code when it is running,  under webpack the module ID is just a number, it is not the actual path to the module as one might expect.

**Comment by GeorgeKnap** [comment]
Also getting this error using webpack + webpack-dev-server setup as described on angular.io pages.

I have tried removing the moduleId assignment on all of my components like this:
`@Component({
    selector: 'my-app',
    templateUrl: '../app.component',
    styleUrls: ['../styles/app.component']
})`
I have also tried to cast it to string since match method works on strings:
`@Component({
    moduleId: module.id.toString(),
    selector: 'my-app',
    templateUrl: '../app.component',
    styleUrls: ['../styles/app.component']
})`
None of these works because the method gets called somewhere internally by webpack or angular and passing anID number.
In my case it's uri 511
`
function _split(uri) {
    return uri.match(_splitRe);
}`

RC4 worked ok, RC5 breaks it.

**Comment by mhamel06** [comment]
I think this will be most common with third party libs that haven't removed all module IDs yet. In my case it was from the new NgModule version of material2.  https://github.com/angular/material2/blob/master/src/components/sidenav/sidenav.ts#L225

My PR should mute the error but something like this would be best to add to the webpack.config as well, to make sure all module.ids are stripped out:

https://github.com/AngularClass/angular2-webpack-starter/issues/843#issuecomment-238765879

**Comment by Daeluse** [comment]
Would a resolution to this be pushed out as a Hotfix, or would it be reserved until RC6? Our current environment rebuilds the node_modules folder regularly, so editing the angular files manually is not an option. The new release brings a lot of desired functionality to the table, and we want to keep our project up to date with it.

@mhamel06 Do you happen to know of something that can be added to the webpack config to temporarily patch this issue?

**Comment by mhamel06** [comment]
I am planning to move this into my local version of angular-cli for now but expect it will likely make its way into the official version soon.

https://github.com/AngularClass/angular2-seed/commit/84aacc18682bbe38631cd3c457d9e90944d6fd0e

**Comment by kucharzyk** [comment]
I hope this will be fixed as patch to rc.5

**Comment by hstarorg** [comment]
I fixed it by:
```javascript
function _split(uri) {
    uri += '';
    return uri.match(_splitRe);
}
```

**Comment by lazyspark** [comment]
I did not face this issue in system js. But with webpack as mentioned above. The webpack introduction page - https://angular.io/docs/ts/latest/guide/webpack.html is not using moduleId. Except of the cheatsheet documentation there is no mention for moduleId. But if I remove module Id from an app using system js it is unable to locate the html components. So seems like it is required.

**Comment by robinkedia** [comment]
Even though I removed moduleId from all my components. I see WebPack is explicitly adding it to dist file.

**Comment by robinkedia** [comment]
https://github.com/angular/angular-cli/issues/1910 - This fix also resolves the issue. (moduleId.toString())

**Comment by vicb** [comment]
should be closed by #11617
